04-cs-fundamentals/classroom/Sorting.md

```diff
@@ -116,35 +177,49 @@ def insertion_sort(array, length)
-&nbsp;&nbsp;&nbsp;&nbsp;[**7, 11, 16, 35, 40, 45, 50, 90, 99**]</br>
-
-<strong>Analysis:</strong> In sorting the most expensive part is a comparison of two elements. Insertion sort algorithm requires
+
+**Example:** Consider the initial unsorted array [99, 45, 35, 40, 16, 50, 11, 7, 90]. Here's what it looks like before and after each iteration of the outer loop. The sorted sub-array is **bolded**
+
+| Iteration | Array |
+|-- |-- |
+| 1. | [**99**, 45, 35, 40, 16, 50, 11, 7, 90] |
+| 2. | [**45, 99**, 35, 40, 16, 50, 11, 7, 90] |
+| 3. | [**35, 45, 99**, 40, 16, 50, 11, 7, 90] |
+| 4. | [**35, 40, 45, 99**, 16, 50, 11, 7, 90] |
+| 5. | [**16, 35, 40, 45, 99**, 50, 11, 7, 90] |
+| 6. | [**16, 35, 40, 45, 50, 99**, 11, 7, 90] |
+| 7. | [**11, 16, 35, 40, 45, 50, 99**, 7, 90] |
+| 8. | [**7, 11, 16, 35, 40, 45, 50, 99**, 90] |
+| 9. | [**7, 11, 16, 35, 40, 45, 50, 90, 99**] |
+
+**Analysis:** In sorting the most expensive part is a comparison of two elements. Insertion sort algorithm requires
+
 - 0 comparisons to insert the first element
 - 1 comparison to insert the second element
 - 2 comparisons to insert the third element
 - ... and so on
 - *n-1* comparisons (worst case) to insert the last element
-Overall, this is *1 + 2 + 3 + ... + (n-1)* = *O(n^2)*
 
-The insertion sort has a complexity of O(n^2). Unlike selection sort, insertion sort has a best case time complexity of O(n) when the items are already sorted. In other words, insertion sort runs in linear time on a nearly sorted list of elements.
+Overall, this is *1 + 2 + 3 + ... + (n-1)* = *O(n<sup>2</sup>)*
+
+The insertion sort has a complexity of O(n<sup>2</sup>). Unlike selection sort, insertion sort has a best case time complexity of O(n) when the items are already sorted. In other words, insertion sort runs in linear time on a nearly sorted list of elements.
+
+## O(n<sup>2</sup>) Algorithms
+
+The sorting algorithms above have a time complexity of O(n<sup>2</sup>).  These algorithms are fine when the array length, *n*, is small, but quickly cease to be useful as the size of the array increases.  They **do** have the advantage of being relatively easy to implement and bubble sort and insertion sort can perform well in the best case, but they do not scale well.  This brings in an overview of our first fairly performant sorting algorithm **Merge Sort**.
+
+## Merge Sort
 
-#### Merge Sort
 Merge sort is a *divide-and-conquer* algorithm. It involves the following three stages:
+
 1. **Divide** the array into two sub-arrays at each step until each sub-array is of size one.
-2. **Sort** each sub-array. (An array of size one is sorted by default.)
-3. **Merge** the sub-arrays into one array by combining two sub-arrays into one at each step.</br></br>
+1. **Sort** each sub-array. (An array of size one is sorted by default.)
+1. **Merge** the sub-arrays into one array by combining two sub-arrays into one at each step.</br></br>
```

Comment:
1, 2, 3 lists make me sad.  I find 1, 2, 3 much easier to read.  (And I'm not sure what our argument for using 1, 1, 1 is other than it's somewhat easier to write.)

Reply:
Just because it's easier to insert content into the middle of lists.  I'd guess this is the reasoning.